openerp/addons/base/ir/ir_qweb/qweb.py

```diff
@@ -0,0 +1,972 @@
+# -*- coding: utf-8 -*-
+
+import ast
+from collections import OrderedDict
+import json
+from urlparse import urlparse
+from lxml import etree, html
+import re
+import itertools
+import textwrap
+import traceback
+
+import openerp.tools
+from odoo.exceptions import QWebException
+
+from openerp.tools import html_escape
+from . import utils
+from .utils import QWebContext, unicodifier
+from .assetsbundle import AssetsBundle
+
+from odoo import models
+
+try:
+    import astor
+except ImportError:
+    astor = None
+
+import logging
+_logger = logging.getLogger(__name__)
+
+
+class QWeb(models.AbstractModel):
+    """ Base QWeb rendering engine
+    * to customize ``t-field`` rendering, subclass ``ir.qweb.field`` and
+      create new models called :samp:`ir.qweb.field.{widget}`
+    * alternatively, override :meth:`~.get_field_for` and return an
+      arbitrary model to use as field converter
+    Beware that if you need extensions or alterations which could be
+    incompatible with other subsystems, you should create a local object
+    inheriting from ``ir.qweb`` and customize that.
+    """
```

Comment:
please update the doc
maybe describe the compilation template -> function
what is the API of function?